pilot/pkg/networking/core/v1alpha3/name_table.go

```diff
@@ -73,16 +78,34 @@ func (configgen *ConfigGeneratorImpl) BuildNameTable(node *model.Proxy, push *mo
 		// IP allocation logic for service entry was unable to allocate an IP.
 		if svcAddress == constants.UnspecifiedIP {
 			// For all k8s headless services, populate the dns table with the endpoint IPs as k8s does.
-			// TODO: Need to have an entry per pod hostname of stateful set but for this, we need to parse
-			// the stateful set object, associate the object with the appropriate kubernetes headless service
-			// and then derive the stable network identities.
+			// And for each individual pod, populate the dns table with the endpoint IP with a manufactured host name from
+			// "statefulset.kubernetes.io/pod-name" that k8s sets.
 			if svc.Attributes.ServiceRegistry == string(serviceregistry.Kubernetes) &&
 				svc.Resolution == model.Passthrough && len(svc.Ports) > 0 {
 				// TODO: this is used in two places now. Needs to be cached as part of the headless service
 				// object to avoid the costly lookup in the registry code
 				for _, instance := range push.ServiceInstancesByPort(svc, svc.Ports[0].Port, nil) {
 					// TODO: should we skip the node's own IP like we do in listener?
 					addressList = append(addressList, instance.Endpoint.Address)
+					if pod, ok := instance.Endpoint.Labels[StatefulSetPodLabel]; ok {
+						address := []string{instance.Endpoint.Address}
+						// Follow k8s naming convention of "pod.svcname.namespace.dnsdomain".
+						// TODO: We should ideally look at pod's DNSDomain. But currently we do not have to
+						// get proxy by IP. We could build that cache in DiscoveryServer or ConfigGenerator.
+						// For now, this serves majority of the use cases.
+						host := pod + "." + svc.Attributes.Name + "." + node.DNSDomain
```

Comment:
won't this be the namespace of the connected proxy, not the real svc?

Reply:
can we do `pod + svc.Hostname`?

Reply:
It has to be service. I will change it to pod+svc.Hostname